File: nbui-core/src/main/java/cn/rubintry/nbui/core/NBUI.java
package cn.rubintry.nbui.core;

import java.util.concurrent.ConcurrentHashMap;

public class NBUI {
    private ConcurrentHashMap<Class<? extends INBUIConfig> , INBUIConfig> uiGlobalConfigMap = new ConcurrentHashMap<>();

    private static volatile NBUI instance;

    public static NBUI getInstance(){
        if(instance == null){
            synchronized (NBUI.class){
                if(instance == null){
                    instance = new NBUI();
                }
            }
        }
        return instance;
    }

    private NBUI(){

    }


    /**
     * 设置UI全局配置，优先级低于直接在view中配置
     */
    public  <T extends INBUIConfig> T config(Class<T> clazz){
        if(uiGlobalConfigMap.containsKey(clazz)){
            return (T) uiGlobalConfigMap.get(clazz);
        }
        try {
            INBUIConfig obj = clazz.newInstance();
            uiGlobalConfigMap.put(clazz , obj);
            return (T) obj;
        }catch (Exception e){

        }
        return null;
    }
}

File: nbui-core/src/main/java/cn/rubintry/nbui/core/NBUI.kt
package cn.rubintry.nbui.core

import java.util.concurrent.ConcurrentHashMap


/**
 * NBUI全局操作类
 */
class NBUI private constructor(){

//    private val uiConfigList = mutableListOf<INBUIConfig>()

    private val uiGlobalConfigMap = ConcurrentHashMap<Class<out INBUIConfig> , INBUIConfig>()

    companion object{
        @Volatile
        private var instance : NBUI ?= null
        @JvmStatic
        fun getInstance(): NBUI {
            if(null == instance){
                synchronized(NBUI::class.java){
                    if(null == instance){
                        instance = NBUI()
                    }
                }
            }
            return instance!!
        }
    }


    /**
     * 设置UI全局配置，优先级低于直接在view中配置
     */
    fun <T : INBUIConfig> config(clazz: Class<T>) : T{
        if(uiGlobalConfigMap.containsKey(clazz)){
            return uiGlobalConfigMap[clazz] as T
        }
        val obj = clazz.newInstance()
        uiGlobalConfigMap[clazz] = obj
        return obj
    }
}